File: evaluations/powermodels/tests/fnm_ingestion/test_g_fnm_1_intermediate_ingestion.jl
#=
Test G-FNM-1: Intermediate Format Ingestion (Two-Check Gate)

Dimension: fnm_ingestion
Network: LARGE (FNM)
Pass condition: (a) PSS/E intermediate CSV loads successfully; (b) record counts match manifest
Tool: PowerModels.jl 0.21.5

PowerModels.jl does NOT support CSV ingestion. It supports MATPOWER .m, PSS/E .raw
(v33), and JSON formats only. The intermediate CSV tables cannot be parsed.

Sub-check (a) confirms the PSS/E intermediate CSV format is unsupported (expected fail).
The MATPOWER fallback path is verified to confirm downstream tests (G-FNM-3/4/5) can proceed.
=#

using PowerModels
using JSON

PowerModels.silence()

function run(;
    csv_dir::String="/workspace/data/fnm/intermediate",
    matpower_fallback::String="/workspace/data/fnm/reference/cleaned/fnm_main_island.m",
)
    results = Dict(
        "status" => "fail",
        "wall_clock_seconds" => 0.0,
        "details" => Dict(),
        "errors" => String[],
        "workarounds" => String[],
    )

    t0 = time()
    try
        # --- Sub-check (a): Attempt CSV ingestion ---
        # PowerModels supports MATPOWER .m, PSS/E .raw, and JSON — NOT CSV.
        # The intermediate CSVs are tabular extracts from PSS/E, not a format
        # PowerModels can parse. This failure is expected and documented.

        csv_attempted = false
        csv_success = false

        if isdir(csv_dir)
            csv_files = filter(f -> endswith(f, ".csv"), readdir(csv_dir))
            if !isempty(csv_files)
                csv_attempted = true
                try
                    test_file = joinpath(csv_dir, csv_files[1])
                    data = PowerModels.parse_file(test_file)
                    csv_success = true
                catch e
                    push!(
                        results["errors"],
                        "CSV ingestion failed (expected): $(typeof(e)): $(sprint(showerror, e))",
                    )
                end
            else
                push!(
                    results["errors"],
                    "No CSV files in intermediate directory ($csv_dir). " *
                    "Intermediate CSVs are NDA-restricted and not committed to version control.",
                )
            end
        else
            push!(results["errors"], "Intermediate CSV directory not found: $csv_dir")
        end

        results["details"]["csv_attempted"] = csv_attempted
        results["details"]["csv_success"] = csv_success
        results["details"]["csv_format_unsupported"] = true
        results["details"]["failure_reason"] = "psse_parse_error"
        results["details"]["ingestion_path"] = "matpower_fallback"

        # --- MATPOWER fallback verification ---
        # G-FNM-1 fails because CSV is unsupported, but we verify the MATPOWER
        # fallback loads correctly so downstream G-FNM-3/4/5 can proceed.
        matpower_success = false

        if isfile(matpower_fallback)
            println("Verifying MATPOWER fallback: $matpower_fallback")
            t_load = time()
            try
                matpower_data = PowerModels.parse_file(matpower_fallback)
                matpower_success = true
                load_time = time() - t_load

                n_bus = length(matpower_data["bus"])
                n_branch = length(matpower_data["branch"])
                n_gen = length(matpower_data["gen"])
                n_load = length(matpower_data["load"])

                slack_buses = [
                    parse(Int, id) for (id, bus) in matpower_data["bus"] if bus["bus_type"] == 3
                ]

                baseMVA = get(matpower_data, "baseMVA", nothing)

                # Check tap ratio preservation
                tap_zero_count = 0
                tap_unity_count = 0
                for (id, br) in matpower_data["branch"]
                    tap = get(br, "tap", 1.0)
                    if tap == 0.0
                        tap_zero_count += 1
                    elseif tap == 1.0
                        tap_unity_count += 1
                    end
                end

                results["details"]["matpower_fallback"] = Dict(
                    "path" => matpower_fallback,
                    "load_time_seconds" => load_time,
                    "bus_count" => n_bus,
                    "branch_count" => n_branch,
                    "gen_count" => n_gen,
                    "load_count" => n_load,
                    "slack_buses" => slack_buses,
                    "baseMVA" => baseMVA,
                    "tap_zero_count" => tap_zero_count,
                    "tap_unity_count" => tap_unity_count,
                )

                println("  Buses:      $n_bus")
                println("  Branches:   $n_branch")
                println("  Generators: $n_gen")
                println("  Loads:      $n_load")
                println("  Slack buses: $slack_buses")
                println("  baseMVA:    $baseMVA")
                println("  Load time:  $(round(load_time, digits=2))s")
            catch e
                push!(
                    results["errors"],
                    "MATPOWER fallback failed: $(typeof(e)): $(sprint(showerror, e))",
                )
            end
        else
            push!(results["errors"], "MATPOWER fallback file not found: $matpower_fallback")
        end

        results["details"]["matpower_fallback_success"] = matpower_success

        # G-FNM-1 status is always fail: the PSS/E intermediate CSV format is
        # not supported by PowerModels. The MATPOWER fallback verification is
        # informational — it enables G-FNM-3/4/5 but does not change G-FNM-1 outcome.
        results["status"] = "fail"

        push!(
            results["workarounds"],
            "PowerModels cannot ingest PSS/E intermediate CSV format. " *
            "MATPOWER .m fallback required for all FNM analysis.",
        )

    catch e
        push!(results["errors"], "$(typeof(e)): $(sprint(showerror, e))")
    finally
        results["wall_clock_seconds"] = time() - t0
    end

    return results
end

# Run and print when executed directly
if abspath(PROGRAM_FILE) == @__FILE__
    result = run()
    println("\n=== RESULTS ===")
    println(JSON.json(result, 2))
end

File: evaluations/pandapower/tests/fnm_ingestion/test_g_fnm_1_intermediate_ingestion.py
"""
Test G-FNM-1: Intermediate Format Ingestion (Two-Check Gate)

Dimension: fnm_ingestion
Network: LARGE (FNM Annual S01)
Pass condition: Sub-check (a) PSS/E intermediate CSV tables load successfully;
                Sub-check (b) post-ingestion fidelity checks pass.
Tool: pandapower 3.4.0

pandapower has no native PSS/E CSV parser. Sub-check (a) attempts to load the
intermediate CSV tables from data/fnm/intermediate/ and is expected to fail.
Sub-check (b) is skipped because (a) fails.
"""

from __future__ import annotations

import json
import os
import time
import traceback


def run(
    intermediate_dir: str = "data/fnm/intermediate",
    manifest_file: str = "data/fnm/manifest.json",
) -> dict:
    """Execute the G-FNM-1 intermediate format ingestion test.

    Sub-check (a): Attempt to load intermediate CSV tables (bus.csv, branch.csv,
    transformer.csv, generator.csv, load.csv, etc.) into pandapower's data model.
    pandapower has no PSS/E CSV import capability, so this is expected to fail.

    Sub-check (b): Post-ingestion fidelity checks (bus count, branch count,
    transformer count, baseMVA, slack bus, tap ratio preservation). Only executed
    if sub-check (a) succeeds.

    Returns:
        dict with keys: status, wall_clock_seconds, details, errors, workarounds
    """
    results = {
        "status": "fail",
        "wall_clock_seconds": 0.0,
        "details": {},
        "errors": [],
        "workarounds": [],
    }

    start = time.perf_counter()
    try:
        import pandapower as pp

        results["details"]["pandapower_version"] = pp.__version__

        # Load manifest for expected record counts
        with open(manifest_file) as f:
            json.load(f)  # validate manifest is parseable

        results["details"]["manifest_loaded"] = True

        # --- Sub-check (a): PSS/E intermediate CSV ingestion ---
        # pandapower does not have a native PSS/E CSV parser.
        # It supports MATPOWER .m/.mat import (from_mpc, from_ppc) and its own
        # JSON/pickle format, but cannot directly parse PSS/E v31 record-type CSVs.

        # Check if intermediate CSV files exist
        expected_tables = [
            "bus",
            "load",
            "fixed_shunt",
            "generator",
            "branch",
            "transformer",
            "area",
            "two_terminal_dc",
            "vsc_dc",
            "impedance_correction",
            "multi_terminal_dc",
            "multi_section_line",
            "zone",
            "interarea_transfer",
            "owner",
            "facts",
            "switched_shunt",
        ]

        csv_files_found = {}
        for table in expected_tables:
            csv_path = os.path.join(intermediate_dir, f"{table}.csv")
            csv_files_found[table] = os.path.exists(csv_path)

        results["details"]["csv_files_found"] = csv_files_found
        # Attempt to find any pandapower API that could parse PSS/E CSV data
        psse_parse_apis = []
        for attr_name in dir(pp):
            attr_lower = attr_name.lower()
            if any(kw in attr_lower for kw in ["psse", "pss_e", "raw", "csv_import"]):
                psse_parse_apis.append(attr_name)

        # Check converter submodule
        try:
            import pandapower.converter as ppc

            for attr_name in dir(ppc):
                attr_lower = attr_name.lower()
                if any(kw in attr_lower for kw in ["psse", "pss_e", "raw"]):
                    psse_parse_apis.append(f"converter.{attr_name}")
        except ImportError:
            pass

        results["details"]["psse_parse_apis_found"] = psse_parse_apis
        results["details"]["subcheck_a"] = {
            "description": "PSS/E intermediate CSV ingestion",
            "result": "fail",
            "reason": "psse_parse_error",
            "explanation": (
                "pandapower has no native PSS/E CSV parser. The tool supports "
                "MATPOWER .m/.mat import (from_mpc, from_ppc) and pandapower's own "
                "JSON/pickle serialization format, but cannot directly ingest PSS/E "
                "v31 record-type CSV tables. No function matching PSS/E import was "
                f"found in the public API. APIs scanned: {psse_parse_apis or 'none'}."
            ),
        }

        results["details"]["ingestion_path"] = "matpower_fallback"

        # Sub-check (a) fails -> sub-check (b) is skipped
        results["details"]["subcheck_b"] = {
            "description": "Post-ingestion fidelity checks",
            "result": "skip",
            "reason": "Sub-check (a) failed; no network model to validate.",
        }

        # Record failure
        results["status"] = "fail"
        results["errors"].append(
            "Sub-check (a) FAIL: pandapower has no PSS/E CSV parser. "
            "Cannot ingest intermediate format CSV tables directly. "
            "Fallback to MATPOWER import path is available for G-FNM-3/4/5."
        )

    except Exception as e:
        results["errors"].append(f"{type(e).__name__}: {e}")
        results["details"]["traceback"] = traceback.format_exc()
    finally:
        results["wall_clock_seconds"] = time.perf_counter() - start

    return results


if __name__ == "__main__":
    result = run()
    print(json.dumps(result, indent=2, default=str))
